Add type hints to BatchSampler

nandan2003 opened the pull request:
Fixes #167670

## Summary
This PR adds missing type annotations to the `__init__` method of `BatchSampler` in `torch/utils/data/sampler.py`.

This aligns the code with modern PyTorch typing standards and helps static analysis tools correctly identify the arguments (`sampler`, `batch_size`, `drop_last`).

## Test Plan
Verified locally using `mypy` and `flake8` to ensure no regressions or linting errors.

**Command:**
```bash
mypy torch/utils/data/sampler.py --ignore-missing-imports --follow-imports=skip
```

**Output:**
```
Success: no issues found in 1 source file
```

eshanhasitha reviewed (commented):
✅

aelavender reviewed (changes requested):
Hi @nandan2003, the code is correct as-is.  From Python 3.10 the `|` syntax denotes unions.  3.10 is now the minimum Python version supported by pytorch.

guilhermeleobas committed:
Add `torch.Tensor.__annotate__` to the `testing_ignore` list (#169013)

Skip `torch.Tensor.__annotate__` when testing for `__torch_override__`
Pull Request resolved: https://github.com/pytorch/pytorch/pull/169013
Approved by: https://github.com/rtimpe, https://github.com/williamwen42
